### PR TITLE
Tdkn 55 improve exceptions

### PR DESCRIPTION
**Pull request by sgandon**
**Please check if the PR fulfills these requirements**

- [ ] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [x] Docs have been added / updated (for bug fixes / features) ?
- [ ] The code coverage on new code >75%
- [ ] The new code does not introduce new technical issues (sonar / eslint)

**What kind of change does this PR introduce?**

- [x] Feature

**What is the current behavior?** (You can also link to an open issue here)
to create a complete exception you need to do this
```
        TalendRuntimeException exception = new TalendRuntimeException(
                UNEXPECTED_ARGUMENT, rootCauseException,
                ExceptionContext.build().put("argument", "Value 1").put("value", "123");

```

**What is the new behavior?**
A new exception builder is introduced to make the wrtting of TalendRuntimeException simpler.
```
        TalendRuntimeException exception = TalendRuntimeException.build(
                UNEXPECTED_ARGUMENT, rootCauseException).set("Value 1", "123");

```
put can still be used but this is not compulsory. This avoid replicating the key everywhere and make is simpler to write exception, I hope.

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/111/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/111/)

File: daikon/src/main/java/org/talend/daikon/exception/ExceptionContext.java [hunk 1]
```diff
@@ -13,21 +13,26 @@
...
  * Additional context for {@link TalendRuntimeException}s when they occur, in the form of key/value pairs.
  */
 public class ExceptionContext implements Serializable {
 
     private static final long serialVersionUID = -8905451634883948364L;
 
-    /** The internal context. */
-    private Map<String, Object> context;
+    /**
+     * The internal context. is a LinkedHashMap to garanty the order of the context entries to help the exception builder. see
```
Comment (RyanSkraba) [inline daikon/src/main/java/org/talend/daikon/exception/ExceptionContext.java]: guarantee

File: daikon/src/test/java/org/talend/daikon/exception/json/JsonErrorCodeTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import org.talend.daikon.exception.ExceptionContext;
+import org.talend.daikon.exception.TalendRuntimeException;
+import org.talend.daikon.exception.error.CommonErrorCodes;
+import org.talend.daikon.exception.error.ErrorCode;
+
+public class JsonErrorCodeTest {
+
+    @Test
+    public void testDeserialize() throws IOException {
+        // check that the CodeError serilization can be deserialized into an JsonCodeError
```
Comment (RyanSkraba) [inline daikon/src/test/java/org/talend/daikon/exception/json/JsonErrorCodeTest.java]: serialization

File: daikon/src/main/java/org/talend/daikon/exception/TalendRuntimeException.java [hunk 3]
```diff
@@ -187,4 +190,98 @@ public Throwable getCause() {
...
+            this.errorCode = code;
+        }
+
+        /**
+         * add new key,value to the context of the exception that will be created
+         */
+        public TalendRuntimeExceptionBuilder put(String key, String value) {
+            if (ecb.getContextSize() != errorCode.getExpectedContextEntries().size()) {
+                ecb.put(key, value);
+            } // or else context is already set so ingor this
```
Comment (RyanSkraba) [inline daikon/src/main/java/org/talend/daikon/exception/TalendRuntimeException.java]: ignore

**Pull request merged**